src/compiler/checker.ts

```diff
@@ -14977,6 +14989,50 @@ namespace ts {
             return strictNullChecks ? getGlobalNonNullableTypeInstantiation(type) : type;
         }
 
+        function addOptionalTypeMarker(type: Type) {
+            return strictNullChecks ? getUnionType([type, optionalType]) : type;
+        }
+
+        function removeOptionalTypeMarker(type: Type): Type {
+            return strictNullChecks ? filterType(type, t => t !== optionalType) : type;
+        }
+
+        function propagateOptionalTypeMarker(type: Type, wasOptional: boolean) {
+            return wasOptional ? addOptionalTypeMarker(type) : type;
+        }
+
+        function checkOptionalExpression(
+            parent: PropertyAccessExpression | QualifiedName | ElementAccessExpression | CallExpression,

```

Comment:
Can you give an example of how `parent` could be a `QualifiedName`?

Reply:
This is called from several places, including `checkPropertyAccessExpressionOrQualifiedName`. You can't actually have a `?.` in a qualified name, but we use the same code path.